src/pairwise.jl

```diff
@@ -307,34 +314,130 @@ julia> pairwise(cor, eachcol(y), skipmissing=:pairwise)
+    end
+
+    # Computing integer indices once for all vectors is faster
+    nminds′ = findall(nminds)
+    # TODO: check whether wrapping views in a custom array type which asserts
+    # that entries cannot be `missing` (similar to `skipmissing`)
+    # could offer better performance
 
-pairwise!(::typeof(cov), dest::AbstractMatrix, x, y;
-          symmetric::Bool=false, skipmissing::Symbol=:none) =
-    pairwise!(_cov, dest, x, y, symmetric=symmetric, skipmissing=skipmissing)
+    x′ = [disallowmissing(view(xi, nminds′)) for xi in x]
+    if x === y
+        return x′, x′
+    else
+        y′ = [disallowmissing(view(yi, nminds′)) for yi in y]
+        return x′, y′
+    end
+end
 
-pairwise(::typeof(cov), x, y; symmetric::Bool=false, skipmissing::Symbol=:none) =
-    pairwise(_cov, x, y, symmetric=symmetric, skipmissing=skipmissing)
+"""
+    handle_pairwise(x::AbstractVector, y::AbstractVector;
+    scratch_fx::AbstractVector=similar(x, nonmissingtype(eltype(x))),
+    scratch_fy::AbstractVector=similar(y, nonmissingtype(eltype(y))))
 
-pairwise!(::typeof(cov), dest::AbstractMatrix, x;
-         symmetric::Bool=true, skipmissing::Symbol=:none) =
-    pairwise!(_cov, dest, x, x, symmetric=symmetric, skipmissing=skipmissing)
+Return a pair `(a,b)`, filtered copies of `(x,y)`, in which elements `x[i]` and
+`y[i]` are excluded if  `ismissing(x[i])||ismissing(y[i])`.
+"""
+function handle_pairwise(x::AbstractVector, y::AbstractVector;
+    scratch_fx::AbstractVector=similar(x, nonmissingtype(eltype(x))),
+    scratch_fy::AbstractVector=similar(y, nonmissingtype(eltype(y))))
+
+    axes(x) == axes(y) || throw(DimensionMismatch("x and y have inconsistent dimensions"))
+    lb = first(axes(x, 1))
+    j = lb - 1
+    @inbounds for i in eachindex(x)
+        if !(ismissing(x[i]) || ismissing(y[i]))
+            j += 1
+            scratch_fx[j] = x[i]
+            scratch_fy[j] = y[i]
+        end
+    end
+
+    return view(scratch_fx, lb:j), view(scratch_fy, lb:j)
+end
 
-pairwise(::typeof(cov), x; symmetric::Bool=true, skipmissing::Symbol=:none) =
-    pairwise(_cov, x, x, symmetric=symmetric, skipmissing=skipmissing)
+#=Condition a) makes equal_sum_subsets useful for load-balancing the multi-threaded section
+of _pairwise! in the non-symmetric case, and condition b) for the symmetric case.=#
+"""
+    equal_sum_subsets(n::Int, num_subsets::Int)::Vector{Vector{Int}}
+
+Divide the integers 1:n into a number of subsets such that a) each subset has
+(approximately) the same number of elements; and b) the sum of the elements in each subset
+is nearly equal. If `n` is a multiple of `2 * num_subsets` both conditions hold exactly.
+
+## Example
+```julia-repl
+julia> StatsBase.equal_sum_subsets(30,5)
+5-element Vector{Vector{Int64}}:
+ [30, 21, 20, 11, 10, 1]
+ [29, 22, 19, 12, 9, 2]
+ [28, 23, 18, 13, 8, 3]
+ [27, 24, 17, 14, 7, 4]
+ [26, 25, 16, 15, 6, 5]
+```
+"""
+function equal_sum_subsets(n::Int, num_subsets::Int)::Vector{Vector{Int}}
```

Comment:
In case this function is kept (see comments above), is it necessary to return a vector of vectors or could one return an iterator of vectors instead?

Reply:
I'm reasonably sure that `equal_sum_subsets` works well in the sense of doing a good job of load balancing. The only reason it currently returns a vector of vectors is that's what I know how to do, so I would need to read up on iterators.

Reply:
In commit a1d9ae9 I have replaced the function `equal_sum_subsets` with an iterator `EqualSumSubsets` whose element type is another iterator `TwoStepRange`. It seems to work well and makes a small reduction in allocations. But please review as this is the first time I've written code to create an iterator.

Reply:
Ah, didn't work on 32 bit. Maybe I need to replace use of Int64 with Int?